Add support for OSC 52 (copy-to-clipboard)

uzxmx opened the pull request:
With this commit, terminal will be able to copy text to the system
clipboard by using OSC 52 MANIPULATE SELECTION DAATA.

We chose not to implement the clipboard querying functionality offered
by OSC 52, as sending the clipboard text to an application without the
user's knowledge or consent is an immense security hole.

We do not currently support the clipboard specifier Pc to specify which
clipboard buffer should be filled

# Base64 encoded `foo`
$ echo -en "\e]52;;Zm9v\a"

# Multiple lines
# Base64 encoded `foo\r\nbar`
$ echo -en "\e]52;;Zm9vDQpiYXI=\a"

Closes #2946.

hh commented:
/retest

uzxmx commented:
Don't know why the CI failed. The error is:

```
src\renderer\dx\BoxDrawingEffect.h#L10
src\renderer\dx\BoxDrawingEffect.h(10,10): Error C1083: Cannot open include file: 'IBoxDrawingEffect_h.h': No such file or directory (compiling source file ..\Window.cpp)
```

But I didn't touch `BoxDrawingEffect.h`. It seems that `BoxDrawingEffect.h` was introduced serveral days ago by 70867df077e03a8ac5d4b230e28d64f90e08045e, which depends on the auto-generated  `IBoxDrawingEffect_h.h` (ignored by git). On my local machine, `IBoxDrawingEffect_h.h` is generated.

jmhbnz commented:
Completed a quick manual test to check the proposed copy to clipboard functionality in relation to my needs. Works exactly as expected :) 

I've been looking forward to this functionality for quite a while, really hope it can make it into 1.0.

Being new to this project here are the steps I followed in case anyone wants to verify/reproduce my manual test:
- Clone `https://github.com/uzxmx/terminal`
- Clone submodules: `git submodule update --init --recursive`
- Built with powershell `Import-Module .\tools\OpenConsole.psm1` followed by `Set-MsBuildDevEnvironment` then `Invoke-OpenConsoleBuild`
- Opened Visual Studio 2019, set build mode for CascadiaPackage project to `Release` and `x64`
- Started the project then in the ensuing terminal used the `echo -en "\e]52;foo\a"` examples outlined in the PR to confirm expected text is placed on the standard windows clipboard.

hh commented:
I haven't been able to test yet, but I think the big target software here is:
- Terminal Muxers 
  - [ ] [tmux](https://github.com/tmux/tmux/wiki/Clipboard#the-clipboard)
  - [ ] [tmate](http://tmate.io) (a fork of tmux with selfhosted shared web + ssh host) 
  - [ ] [mosh]()

- Editors
  - [ ] vim
  - [ ] emacs

Chromium has an folder with osc52 scripts for both editors and a quick shell script to test under tmux/tmate: https://github.com/chromium/hterm/tree/master/etc

This gist may help in configuring tmux/tmate:

uzxmx commented:
@hh  Agree with you. The reason I made this PR is that I really need to rely on this feature when I ssh into a vagrant box.

Considering you mentioned tmux and vim, I also want to share some of my experiences. Actually without OSC 52 support, I've already been able to copy to and paste from windows system clipboard when using tmux and neovim. 

My environment is:  Windows Terminal + Ubunt WSL + Nvim (only in WSL enough) + Tmux.

The core principle is to pipe text to `clip.exe` (provided by Windows system) and get clipboard contents by a get-from-clipboard executable (Windows system hasn't provided) or some command from powershell. It works the same way as pbcopy/pbpaste in OSX system.

Unlike vim, Nvim provides clipboard providers, so we can easily add a custom provider. Here are the [reference](https://github.com/uzxmx/dotfiles/blob/master/vim/plugin/wsl.vim) for nvim and [reference](https://github.com/uzxmx/dotfiles/blob/1f2350899000812384472a955a9d049f7a56f47f/tmux.conf#L92-L96) for tmux configuration in my dotfiles repo. This [article](https://uzxmx.github.io/wsl-tricks.html#access-windows-clipboard-in-nvim-on-wsl) gives more details on that and it also summarizes serveral WSL tricks. 

But without  OSC 52 support, it's impossible to do that on a remote machine.

nicm commented:
This is incorrect:
~~~~
echo -en "\e]52;foo\a"
~~~~
1) OSC 52 takes two arguments, the first of which is the clipboards to set (empty default).

2) The text must be Base64-encoded.

Take a look at: https://invisible-island.net/xterm/ctlseqs/ctlseqs.html

~~~~
Ps = 5 2  ⇒  Manipulate Selection Data.  These controls may
          be disabled using the allowWindowOps resource.  The parameter
          Pt is parsed as
               Pc ; Pd
          The first, Pc, may contain zero or more characters from the
          set c , p , q , s , 0 , 1 , 2 , 3 , 4 , 5 , 6 , and 7 .  It is
          used to construct a list of selection parameters for clip-
          board, primary, secondary, select, or cut buffers 0 through 7
          respectively, in the order given.  If the parameter is empty,
          xterm uses s 0 , to specify the configurable primary/clipboard
          selection and cut buffer 0.

          The second parameter, Pd, gives the selection data.  Normally
          this is a string encoded in base64 (RFC-4648).  The data
          becomes the new selection, which is then available for pasting
          by other applications.

zadjii-msft commented on src/terminal/adapter/adaptDispatch.cpp:
```diff
@@ -2070,6 +2070,17 @@ bool AdaptDispatch::SetCursorColor(const COLORREF cursorColor)
+// - True if handled successfully. False otherwise.
+bool AdaptDispatch::CopyToClipboard(const std::wstring_view content)
+{
+    return _pConApi->CopyToClipboard(content);
```
You could probably avoid a lot of plumbing here by just returning false here. When an `AdaptDispatch` method returns false, and the console is working in conpty mode, we'll automatically try to pass the string through to the connected terminal application. That way, you wouldn't need to manually plumb the sequence all the way through `VtIo` into the `VtEngine`, etc 😄

uzxmx replied:
I missed this point. That's much simpler.

zadjii-msft commented on src/terminal/parser/OutputStateMachineEngine.cpp:
```diff
@@ -744,6 +745,9 @@ bool OutputStateMachineEngine::ActionOscDispatch(const wchar_t /*wch*/,
         success = _GetOscSetColor(string, color);
         break;
+    case OscActionCodes::CopyToClipboard:
+        success = _GetOscString(string, copyContent);
```
As @nicm [called out](https://github.com/microsoft/terminal/pull/5823#issuecomment-626568429), we should make sure we're parsing this the same way it's defined in the standard.

uzxmx replied:
Sure, I've made this change.

zadjii-msft commented on src/terminal/parser/OutputStateMachineEngine.hpp:
```diff
@@ -136,6 +136,7 @@ namespace Microsoft::Console::VirtualTerminal
             SetForegroundColor = 10,
```
Any reason this file and the `cpp` for this class had the execute bits set on it? I'm sure it doesn't _actually matter, but it's certainly curious...

uzxmx replied:
I don't actually get your point here. Do you mean the enum element `CopyToClipboard` is not a proper name, and should be changed to `SetXXX`?

zadjii-msft replied:
Oh no I was talking about this:

![image](https://user-images.githubusercontent.com/18356694/81978342-00b8f600-95f1-11ea-977e-c42496adfe33.png)

I've never seen anything like `100644 → 100755` before, but they do look an awful lot like `chmod` bits

DHowett-MSFT replied:
This can be fixed with...

`git add --chmod -x xyz/OutputStateMachineEngine.hpp`

(once for every file that is changed, of course.)

uzxmx replied:
@DHowett-MSFT Thanks. @zadjii-msft Got what you meant.

That may be because I used vim to edit that file through terminal and some unexpected thing happened. Already fixed.

zadjii-msft commented on src/terminal/parser/stateMachine.cpp:
```diff
@@ -254,7 +254,7 @@ static constexpr bool _isOscTerminationInitiator(const wchar_t wch) noexcept
 static constexpr bool _isOscInvalid(const wchar_t wch) noexcept
 {
-    return wch <= L'\x17' ||
+    return (wch <= L'\x17' && wch != L'\x0a' && wch != L'\x0d') ||
```
Isn't this redundant?

uzxmx replied:
Actually, this wasn't. Before I added base64 support, to support passing "\r\n" into the clipboard, the state machine should view `\r 0x0d` and `\n 0x0a` as valid OSC characters. But `wch <= L'\x17'` will make them invalid, so I changed it to `(wch <= L'\x17' && wch != L'\x0a' && wch != L'\x0d')`.

Anyway, that change is no longer needed, because base64-encoded string won't include `\r` and `\n`.

uzxmx commented:
@nicm
 
Thanks for pointing out that information. Really helpful. I referenced tmux's OSC 52 implementation, and finished adding base64 mechanism to make it conform to the standard.

@zadjii-msft

Changes have been made. Please have a look. Currently, the first parameter `Pc` is ignored. So it always uses the system clipboard. 

When the parameter is neither a valid base64 string nor `?`, the standard says `the selection is cleared`. So would you suggest we clear the content of the system clipboard?

Another question is should we implement below in this PR? Actually I tried to experiment, but have no success. I didn't find the event hook when a newline for shell prompt e.g. `>` appears, or something that's called ready to accept user input, or some function to feed some string after a new prompt. If this feature is required, could you give me some guidance?

```
If the second parameter is a ? , xterm replies to the host
with the selection data encoded using the same protocol.  It
uses the first selection found by asking successively for each
item from the list of selection parameters.
```

uzxmx commented on src/terminal/parser/OutputStateMachineEngine.cpp:
```diff
@@ -1739,6 +1748,47 @@ bool OutputStateMachineEngine::_GetRepeatCount(std::basic_string_view<size_t> pa
+                content = reader.ReadString(buffer.Length());
+                success = true;
+            }
+            catch (...)
```
@zadjii-msft  I wanted to use CATCH_LOG here, but when an invalid base64 comes, it will freeze the terminal and make it unresponsive. I don't know the exact reason, maybe WIL cannot parse exceptions from `winrt::Windows::Security::Cryptography` well?

zadjii-msft replied:
Huh, that's definitely weird...

I bet it's because the winrt header that defines `winrt::Windows::Security::Cryptography` is getting included _after_ WIL - IIRC, the order matters for these things. Maybe if you move the header to before `wil` in `precomp`, it might work?

I'm totally spitballing here

uzxmx replied:
Tried with no luck. :(

zadjii-msft replied:
I figured this out! If you include `<wil/cppwinrt.h>` before the other `winrt` headers, you can use `CATCH_LOG` here! There's a comment with more details in the `precomp.h` file.

DHowett-MSFT commented:
So, the question about supporting clipboard _query_: I don't think we should. It's generally regarded as a security boundary issue if an application, especially a _remote_ one, can silently request a read from your clipboard. We'd need to build up some infrastructure to make sure that that operation is permitted before we add support for it! 

It's very good that this pull request only adds the clipboard overwriting half :smile:

jsoref commented on src/terminal/parser/ut_parser/OutputEngineTest.cpp:
```diff
@@ -2011,4 +2018,63 @@ class StateMachineExternalTest final
+
+        pDispatch->ClearState();
+
+        mach.ProcessString(L"\x1b]52;;Zm9vDQpiYXI=\x07");
```
The bot doesn't like this line.

You should be able to add something like this to https://github.com/microsoft/terminal/blob/master/.github/actions/spell-check/patterns/patterns.txt: 
`\bmach\.ProcessString\(L"\x1b]52;;Zm9vDQpiYXI=\x07"\)`

(A future version of the bot should actually identify this line...)

uzxmx replied:
Fixed. Thanks!

uzxmx commented:
@zadjii-msft  I think I've finished the requested changes. Is there any other change that needs to be made?

zadjii-msft commented:
Hey sorry for the delay! The whole team was a bit randomized for the last week due to Build and Memorial Day (here in the US). I'll look over this _right now_ ☺️

zadjii-msft commented on src/terminal/parser/precomp.h:
```diff
@@ -10,6 +10,9 @@ Module Name:
 - Avoid including internal project headers. Instead include them only in the classes that need them (helps with test project building).
 */
 
+#include <winrt/Windows.Security.Cryptography.h>
```
Okay I figured it out!

You need to include `<wil/cppwinrt.h>` _after_ `wil`, but before the rest of the winrt headers. So change this file to the following:

```c++
#include "LibraryIncludes.h"

#include <windows.h>

#include <sal.h>

#include <cstdio>

#define ENABLE_INTSAFE_SIGNED_FUNCTIONS
#include <intsafe.h>

#include "telemetry.hpp"
#include "tracing.hpp"

#include <wil/cppwinrt.h>
#include <winrt/Windows.Security.Cryptography.h>
#include <winrt/Windows.Storage.Streams.h>

```

uzxmx replied:
Works like a charm. Cheers!

zadjii-msft commented on src/terminal/parser/ut_parser/OutputEngineTest.cpp:
```diff
@@ -2011,4 +2018,63 @@ class StateMachineExternalTest final
+
+        pDispatch->ClearState();
+
+        mach.ProcessString(L"\x1b]52;s0;Zm9v\x07");
```
nit: I might add comments explaining what each of these cases is - it's hard to see at a glance why `\x1b]52;s0;Zm9v\x07` should work and `\x1b]52;Zm9v\x07` doesn't, without having to refer to the spec.

uzxmx replied:
Explanations have been added to the tests.

zadjii-msft commented:
@msftbot make sure @DHowett signs off on this

DHowett commented:
Sorry, I’m running behind on my mails. Quick concern: I’m not sure we should be adding a dependency (even a header dependency) from the shared components (VT parser, for example!) to WinRT things. That code is ingested directly into the version of conhost.exe that ships in Windows!

Sorry if that’s not the case here- I may be looking at an old set of comments in my inbox :) if Mike already addressed this, as well, not a problem!

DHowett commented:
(Conhost is at a lower level than a bunch of WinRT things, and doesn’t have any dependencies other than GDI and COM and stuff. We may want to have a team chat about dependency risk here! It might be fine, but we have to choose to cross that bridge explicitly.)

uzxmx commented:
The WinRT dependency is included to decode base64 string. If that dependency is not allowed because of dependency risk, I may add a custom base64 encode/decode implementation in VT parser. If so, please let me know.

zadjii-msft commented:
I honestly don't know how acceptable that is. I certainly don't mind, it's not like we're adding any of the XAML headers, that would blow up the compilation times. What I'm wary about it making sure this works in the (closed source) Windows build correctly as well. I don't know off the top of my head, @DHowett might but it'll probably be a little while before he's freed up to investigate. Thanks for your patience ☺️

j4james commented:
If you want a Win32 API, you can always use [`CryptStringToBinaryW`](https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-cryptstringtobinaryw). It's part of the crypt32 dll, so I don't know if that's more acceptable? 

Just note that the decoder is fairly strict, so if we want to be XTerm compatible, we'll need to do a bit of preprocessing to remove any invalid chars and pad to an appropriate length (I don't know if the WinRT one is any better in that regard). In the end it might just be easier to roll your own.

j4james commented on src/terminal/parser/OutputStateMachineEngine.hpp:
```diff
@@ -230,6 +231,10 @@ namespace Microsoft::Console::VirtualTerminal
         bool _GetRepeatCount(const std::basic_string_view<size_t> parameters,
                              size_t& repeatCount) const noexcept;
 
+        bool OutputStateMachineEngine::_GetOscSetClipboard(const std::wstring_view string,
```
Is there a reason why the `OutputStateMachineEngine::` class prefix is included here in the header?

uzxmx replied:
Oh, my silly mistake. I copied that directly from the implementation. 😂  Already fixed.

miniksa commented:
> If you want a Win32 API, you can always use [`CryptStringToBinaryW`](https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-cryptstringtobinaryw). It's part of the crypt32 dll, so I don't know if that's more acceptable?
> 
> Just note that the decoder is fairly strict, so if we want to be XTerm compatible, we'll need to do a bit of preprocessing to remove any invalid chars and pad to an appropriate length (I don't know if the WinRT one is any better in that regard). In the end it might just be easier to roll your own.

I did a quick check and `CryptStringToBinaryW` and `Windows::Security::Cryptography::CryptographicBuffer.FromBase64String` end up in different algorithms for decoding the text. I thought it might go to the same place deep under the hood, but it doesn't.

uzxmx commented:
@DHowett @zadjii-msft To move this feature forward, I removed the WinRT dependency and added custom implementation of base64 decoding that conforms to [standard base64](https://tools.ietf.org/html/rfc4648#section-4) with padding properly (for completeness, encoding is also added). The main idea was borrowed from [tmux base64 implementation](https://github.com/tmux/tmux/blob/master/compat/base64.c).

In the future, if we prefer a Windows API for base64, then we can replace in no time.

Please have a review. Thanks!

DHowett commented:
Thanks. Really sorry for the delay on my side :smile:

DHowett commented on src/terminal/adapter/ITermDispatch.hpp:
```diff
@@ -110,6 +110,8 @@ class Microsoft::Console::VirtualTerminal::ITermDispatch
     virtual bool SetCursorStyle(const DispatchTypes::CursorStyle cursorStyle) = 0; // DECSCUSR
     virtual bool SetCursorColor(const COLORREF color) = 0; // OSCSetCursorColor, OSCResetCursorColor
 
+    virtual bool SetClipboard(std::wstring_view content) = 0; // OscSetClipboard
```
```suggestion
    virtual bool SetClipboard(std::wstring_view content) = 0; // OSCSetClipboard
```

for naming consistency (when we inevitably get a tool that parses these comments)

DHowett replied:
(Which is in the works back on the first page of pull requests :P)

uzxmx replied:
Fixed.

DHowett commented on src/terminal/parser/OutputStateMachineEngine.cpp:
```diff
@@ -935,6 +941,13 @@ bool OutputStateMachineEngine::ActionOscDispatch(const wchar_t /*wch*/,
+            {
+                success = _dispatch->SetClipboard(setClipboardContent);
+            }
+            TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCSCC);
```
This may be the wrong TermTelemetry code!

uzxmx replied:
I checked http://inwap.com/pdp10/ansicode.txt to find a proper name for that code, but failed. So I named it as `OSCSCB` where `SCB` means `set clipboard`. I didn't use `SC` because it's too generic and may cause conflict.

DHowett commented on src/terminal/parser/OutputStateMachineEngine.hpp:
```diff
@@ -252,6 +253,10 @@ namespace Microsoft::Console::VirtualTerminal
                              size_t& repeatCount) const noexcept;
 
+        bool _GetOscSetClipboard(const std::wstring_view string,
+                                 std::wstring& title,
```
```suggestion
                                 std::wstring& content,
```

consistency

uzxmx replied:
Fixed.

DHowett commented on src/terminal/parser/base64.cpp:
```diff
@@ -0,0 +1,172 @@
+// - dst - Destination to decode into.
+// Return Value:
+// - true if decoding successfully, otherwise false.
+bool Base64::s_Decode(const std::wstring_view src, std::wstring& dst) noexcept
```
Should we reserve approx. enough space in `dst` so we don't have to keep growing it?

uzxmx replied:
Good point. I've added that. Thanks!

DHowett commented on src/terminal/parser/parser-common.vcxitems:
```diff
@@ -11,6 +11,7 @@
     <ClCompile Include="..\precomp.cpp">
       <PrecompiledHeader>Create</PrecompiledHeader>
     </ClCompile>
+    <ClCompile Include="..\base64.cpp" />
```
You'll need to add `base64.cpp` to the `sources.inc` file for the parser as well :smile: thanks!

(I was going to do it myself when I reflect these changes back into Windows, but I saw that you already updated one of the sources files!)

uzxmx replied:
I had done it, but found without that it also worked, so removed it when committing. Thanks for pointing this out, I've added it back.

zadjii-msft commented on src/terminal/parser/base64.cpp:
```diff
@@ -0,0 +1,186 @@
+
+        switch (state)
+        {
+        case 0:
```
I'd love for these states to be a named enum, but it's okay as-is for now.

zadjii-msft commented on src/terminal/parser/OutputStateMachineEngine.cpp:
```diff
@@ -1859,6 +1872,37 @@ bool OutputStateMachineEngine::_GetRepeatCount(std::basic_string_view<size_t> pa
+                                                   std::wstring& content,
+                                                   bool& queryClipboard) const noexcept
+{
+    const size_t pos = string.find(';');
```
I supposed @DHowett is right, and we could save ourselves a bit of work with a 
```c++
if (_pfnFlushToTerminal != nullptr)
{
    return false;
}
```

here. That _should_ cause conpty to pass this through without decoding it once first. That being said, I'm pretty sure _setting the clipboard_ isn't something that's going to be done in a tight loop or end up as any sort of hot path, so I wouldn't block this PR on account of that.

j4james commented on src/terminal/parser/telemetry.hpp:
```diff
@@ -97,6 +97,7 @@ namespace Microsoft::Console::VirtualTerminal
             OSCFG,
             OSCBG,
             DECALN,
+            OSCSCB,
```
I believe you also need to update the `TermTelemetry` class with the new telemetry entry. See here:
https://github.com/microsoft/terminal/blob/96a77cb74b4ce2a13facf25daf5d8fd6234c19f9/src/terminal/parser/telemetry.cpp#L269-L270

DHowett replied:
Indeed! This is the last thing I'll block on, and then this is totally ready.

uzxmx commented:
@DHowett Updated!
